dev/InfoBar/InfoBar.cpp

```diff
@@ -0,0 +1,230 @@
+            {
+                if (notify && peer)
+                {
+                    auto const notificationString = StringUtil::FormatString(
+                        ResourceAccessor::GetLocalizedStringResource(SR_InfoBarOpenedNotification),
+                        Title().data(),
+                        Message().data());
+
+                    winrt::get_self<InfoBarAutomationPeer>(peer)->RaiseOpenedEvent(Severity(), notificationString);
+                }
+
+                winrt::VisualStateManager::GoToState(*this, L"InfoBarVisible", false);
+                winrt::AutomationProperties::SetAccessibilityView(*this, winrt::AccessibilityView::Control);
+                m_isVisible = true;
+            }
+            else
+            {
+                if (notify && peer)
+                {
+                    auto const notificationString = ResourceAccessor::GetLocalizedStringResource(SR_InfoBarClosedNotification);
+
+                    winrt::get_self<InfoBarAutomationPeer>(peer)->RaiseClosedEvent(Severity(), notificationString);
+                }
+
+                winrt::VisualStateManager::GoToState(*this, L"InfoBarCollapsed", false);
+                winrt::AutomationProperties::SetAccessibilityView(*this, winrt::AccessibilityView::Raw);
+                m_isVisible = false;
+            }
+        }
+    }
+}
+
+void InfoBar::UpdateSeverity()
+{
+    auto severityState = L"Informational";
+
+    switch (Severity())
+    {
+        case winrt::InfoBarSeverity::Success:  severityState = L"Success";  break;
+        case winrt::InfoBarSeverity::Warning:  severityState = L"Warning";  break;
+        case winrt::InfoBarSeverity::Error:    severityState = L"Error"; break;
+    };
+
+    winrt::VisualStateManager::GoToState(*this, severityState, false);
+}
+
+void InfoBar::UpdateIcon()
+{
+    auto const templateSettings = winrt::get_self<::InfoBarTemplateSettings>(TemplateSettings());
+    if (auto const source = IconSource())
+    {
+        templateSettings->IconElement(SharedHelpers::MakeIconElementFrom(source));
+    }
+    else
+    {
+        templateSettings->IconElement(nullptr);
+    }
+}
+
+void InfoBar::UpdateIconVisibility()
+{
+    winrt::VisualStateManager::GoToState(*this, IsIconVisible() ? (IconSource() ? L"UserIconVisible" : L"StandardIconVisible") : L"NoIconVisible", false);
+}
+
+void InfoBar::UpdateCloseButton()
+{
+    winrt::VisualStateManager::GoToState(*this, IsClosable() ? L"CloseButtonVisible" : L"CloseButtonCollapsed", false);
+}
+
+void InfoBar::OnForegroundChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    UpdateForeground();
+}
+
+void InfoBar::UpdateForeground()
+{
+    // If Foreground is set, then change Title and Message Foreground to match.
+    winrt::VisualStateManager::GoToState(*this, ReadLocalValue(winrt::Control::ForegroundProperty()) == winrt::DependencyProperty::UnsetValue() ? L"ForegroundNotSet" : L"ForegroundSet", false);
```

Comment:
@jevansaks and @mikehillberg reacted pretty strongly the last time I tried to use ReadLocalValue().

Reply:
There is simply no other way to do this.

Reply:
Not sure the specific context here, but it's bizarre behavior in general. Setting a property value from code vs with a Style shouldn't change the target's behavior.

Reply:
Wait, is it considered unset if it's set in a style instead of in code? So basically, the problem is that we want Foreground to apply to both Title and Message, but by default Title and Message have different staticresource brushes. I guess I could just listen to when Foreground changes and set the visual state then...? There would be no way to get back to the default behavior, but maybe that's fine.